### PR TITLE
Dockerize Jenkins masters on DCOS

### PR DESCRIPTION
**Pull request by rji**
This PR, for all intents and purposes, rewrites the mesosphere/jenkins-mesos project up to this point. It moves to using Docker for a distribution mechanism, introduces Tomcat to give us greater control over URL rewriting in DCOS, and includes a bootstrap script that modifies specific Jenkins parameters at runtime based on their current Marathon values (e.g. `$HOST`, `$PORT0`, etc).

The bootstrap script is also responsible for populating a host volume (mounted in the container at `/var/jenkins_home`) should this be the first time that Jenkins is coming up. This allows us to use a NFS share mounted on each Mesos/DCOS agent for persisting Jenkins data across instance failover/restart.

See also: https://github.com/mesosphere/multiverse/pull/26

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -1,8 +1,41 @@
...
-RUN apt-key adv --keyserver hkp://keyserver.ubuntu.com:80 --recv-keys 36A1D7869245C8950F966E92D8576A8BA88D21E9
-RUN sh -c "echo deb https://get.docker.io/ubuntu docker main > /etc/apt/sources.list.d/docker.list"
 RUN apt-get update
-RUN apt-get install lxc-docker-1.6.2
-RUN usermod -a -G docker $USER
\ No newline at end of file
+RUN apt-get install -y git python zip
+RUN /usr/local/jenkins/bin/plugin_install.sh "${JENKINS_STAGING}/plugins"
+
+EXPOSE 8080
```
Comment (ssk2) [inline Dockerfile]: Is necessary to expose a port here? Particularly given we're changing the running port later.

File: README.md [hunk 2]
```diff
@@ -1,18 +1,62 @@
...
+## Overview
+This repo contains a [Dockerfile](Dockerfile) that runs Jenkins inside the
+[Apache Tomcat][tomcat-home] Java application server. It also provides several
+Jenkins plugins and a basic Jenkins configuration, to get you up and running
+quickly with Jenkins on the DCOS.
+
+## Included in this repo
+Base packages:
+  * [Apache Tomcat][tomcat-home] 8.0.27
+  * [Jenkins][jenkins-home] 1.609.3 (LTS)
```
Comment (ssk2) [inline README.md]: Worth upgrading to 1.625.1? It looks like this is the newest LTS.

File: scripts/bootstrap.py [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+    """
+    tree = ET.parse(config_xml)
+    root = tree.getroot()
+    mesos = root.find('./clouds/org.jenkinsci.plugins.mesos.MesosCloud')
+
+    mesos_frameworkName = mesos.find('./frameworkName')
+    mesos_frameworkName.text = name
+
+    mesos_jenkinsURL = mesos.find('./jenkinsURL')
+    mesos_jenkinsURL.text = ''.join(['http://', host, ':', port, context])
```
Comment (ssk2) [inline scripts/bootstrap.py]: Will this still append /services/jenkins to the URL?

File: scripts/bootstrap.py [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+        'jenkins.model.JenkinsLocationConfiguration.xml' file
+    :param host: the Mesos agent the task is running on
+    :param port: the Mesos port the task is running on
+    :param context: the context of the Jenkins application, e.g.
+        '/service/jenkins'
+    """
+    tree = ET.parse(location_xml)
+    root = tree.getroot()
+    jenkinsUrl = root.find('jenkinsUrl')
+    jenkinsUrl.text = ''.join(['http://', host, ':', port, context])
```
Comment (ssk2) [inline scripts/bootstrap.py]: Ditto: will this still append /services/jenkins to the URL?

**Comment by ssk2** [comment]
A few comments but I think we're good otherwise! :100:

File: Dockerfile (hunk 1 above)
Comment (rji) [inline Dockerfile]: Ah good catch, this was probably left over from when I was attempting to get this working with bridged networking. I'll go ahead and remove that.

File: README.md (hunk 2 above)
Comment (rji) [inline README.md]: It looks like that was released about a week ago. Let me try it out first, but I suspect that yes, we would want to ship the latest LTS here.

File: scripts/bootstrap.py (hunk 3 above)
Comment (rji) [inline scripts/bootstrap.py]: Good catch, it looks like when I rebased I lost a commit. Let me fix that up quick.

File: scripts/bootstrap.py (hunk 4 above)
Comment (rji) [inline scripts/bootstrap.py]: Good catch, it looks like when I rebased I lost a commit. Let me fix that up quick.

**Comment by rji** [comment]
Hey @ssk2 -- I think I got all your feedback incorporated into the latest revision of the PR. Thanks! Check it out and let me know if I missed anything.

**Comment by ssk2** [comment]
Looks good, thanks!

**Pull request merged**